### PR TITLE
Didtype index

### PR DESCRIPTION
**Pull request by nitro-neal**
**This change creates a new data store did-type-index and on a  DID:ION Create Event it adds this to an indexed table.**

This table can be queried and on receiving a DID type string, will return all DIDs associated with that DID type.

Changes to the codebase include:

- Adding a new didTypeStore
- Adding DidTypeModel {
    didUniqueSuffix: string;
    didType: string;
  }
 - Small updates to transaction processor to store did type
 - Adding didTypeStore to constructor of Request Handler
 - New handleDidTypeRequest function in request handler

TODO:
- copy all of latest/ folder into 1.0/ so that they are exact copies
- sync bitcoin-test net for testing
- sync bitcoin-mainnet for testing

I've created a new Store "DidTypeStore" that behaves similarly to the operation store. This allows did type lookups to happen independently of the operation store processing.
image

### CODE REVIEW

File: sidetree/lib/core/MongoDbDidTypeStore.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export default class MongoDbDidTypeStore extends MongoDbStore implements IDidTypeStore {
+  /** MongoDB collection name under the database where the operations are stored. */
+  public static readonly collectionName: string = 'didtype';
+
+  constructor (serverUrl: string, databaseName: string) {
+    super(serverUrl, MongoDbDidTypeStore.collectionName, databaseName);
+  }
+
+  public async createIndex () {
+    await this.collection.createIndex({ didType: 1 });
```
Comment (decentralgabe) [inline sidetree/lib/core/MongoDbDidTypeStore.ts:18]: what does the value `1` mean here?

File: sidetree/lib/core/models/DidTypeModel.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+/**
+ * Did Type Model
+ */
+export default interface DidTypeModel {
+    /** The DID unique suffix. */
+    didUniqueSuffix: string;
+    /** The didtype. */
```
Comment (decentralgabe) [inline sidetree/lib/core/models/DidTypeModel.ts:8]:
```suggestion
    /** The did type. */
```

File: sidetree/lib/core/versions/1.0/RequestHandler.ts [hunk 3]
```diff
@@ -15,6 +15,8 @@ import Resolver from '../../Resolver';
 import ResponseModel from '../../../common/models/ResponseModel';
 import ResponseStatus from '../../../common/enums/ResponseStatus';
 import SidetreeError from '../../../common/SidetreeError';
+import IDidTypeStore from '../../interfaces/IDidTypeStore';
+// import MongoDbDidTypeStore from '../../MongoDbDidTypeStore';
```
Comment (decentralgabe) [inline sidetree/lib/core/versions/1.0/RequestHandler.ts:19]: can remove?

File: sidetree/lib/core/versions/1.0/TransactionProcessor.ts [hunk 4]
```diff
@@ -31,12 +33,14 @@ export default class TransactionProcessor implements ITransactionProcessor {
...
     private downloadManager: DownloadManager,
     private operationStore: IOperationStore,
+    private didTypeStore : IDidTypeStore,
     private blockchain: IBlockchain,
     private versionMetadataFetcher: IVersionMetadataFetcher) {
   }
 
   public async processTransaction (transaction: TransactionModel): Promise<boolean> {
 
+    console.log('1.1');
```
Comment (decentralgabe) [inline sidetree/lib/core/versions/1.0/TransactionProcessor.ts:43]: what are these logs for?

File: sidetree/lib/core/versions/latest/TransactionProcessor.ts [hunk 5]
```diff
@@ -288,6 +300,25 @@ export default class TransactionProcessor implements ITransactionProcessor {
     return chunkFileModel;
   }
 
+  private composeDidType (coreIndexFile: CoreIndexFile): DidTypeModel[] {
```
Comment (decentralgabe) [inline sidetree/lib/core/versions/latest/TransactionProcessor.ts:303]: can you add a comment to this method?

**Review by decentralgabe — COMMENTED** [review]
I think you need....

1. to add code to add DIDs to the index as they're seen by the node on an ongoing basis
2. a background process like job to populate the index from historical state

these can be the same thing, kind of an 'index propagator' of sorts that keeps the last block number checked, so it can run and populate the index up to the current block.

Would also appreciate a review from @thehenrytsai and @csuwildcat

File: sidetree/lib/core/versions/1.0/TransactionProcessor.ts (hunk 4 above)
Comment (nitro-neal) [reply sidetree/lib/core/versions/1.0/TransactionProcessor.ts:43]: Removed, 
Everything in the 1.0 directory will be a copy of everything I change in latest directory, will update the PR to reflect that

File: sidetree/lib/core/MongoDbDidTypeStore.ts (hunk 1 above)
Comment (nitro-neal) [reply sidetree/lib/core/MongoDbDidTypeStore.ts:18]: The value 1 means the index will be created in ascending order for the didType field.

**Comment by nitro-neal** [comment]
@decentralgabe I added an image at the top which shows where the new DidTypeStore lives. I think that:
1. Is handled because this follows the same path as the operation store and the observer keeps running to update. 
2. I think this is handled, it will propagate the historical state in the exact same way the operation store does it.

Could definitely be wrong, thanks for being thorough !

File: sidetree/lib/core/MongoDbDidTypeStore.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import { Long } from 'mongodb';
+import MongoDbStore from '../common/MongoDbStore';
+
+/**
+ * Implementation of DidTypeStore that stores the didtype data in
+ * a MongoDB database.
+ */
+export default class MongoDbDidTypeStore extends MongoDbStore implements IDidTypeStore {
+  /** MongoDB collection name under the database where the operations are stored. */
+  public static readonly collectionName: string = 'didtype';
```
Comment (thehenrytsai) [inline sidetree/lib/core/MongoDbDidTypeStore.ts:12]: following convention:

```suggestion
  public static readonly collectionName: string = 'did-type';
```

File: sidetree/README.md [hunk 7]
```diff
@@ -17,7 +17,7 @@ See the [API spec](https://identity.foundation/sidetree/api/) for the full API s
 
 Code Coverage
 
-![Statements](https://img.shields.io/badge/statements-100%25-brightgreen.svg?style=flat) ![Branches](https://img.shields.io/badge/branches-100%25-brightgreen.svg?style=flat) ![Functions](https://img.shields.io/badge/functions-100%25-brightgreen.svg?style=flat) ![Lines](https://img.shields.io/badge/lines-100%25-brightgreen.svg?style=flat)
+![Statements](https://img.shields.io/badge/statements-7.6%25-red.svg?style=flat) ![Branches](https://img.shields.io/badge/branches-1.04%25-red.svg?style=flat) ![Functions](https://img.shields.io/badge/functions-2.15%25-red.svg?style=flat) ![Lines](https://img.shields.io/badge/lines-7.71%25-red.svg?style=flat)
```
Comment (thehenrytsai) [inline sidetree/README.md:20]: Should definitely keep coverage at 100%

File: sidetree/lib/core/interfaces/IDidTypeStore.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  insert (didTypes: DidTypeModel[]): Promise<void>;
+
+  /**
+   * Gets all did unique suffix of the given DID type
+   * @param didType the type of the did
+   */
+  get (didType: string): Promise<string[]>;
+
+  /**
+   * Deletes all operations with transaction number greater than the provided parameter.
```
Comment (thehenrytsai) [inline sidetree/lib/core/interfaces/IDidTypeStore.ts:20]:
```suggestion
   * Deletes all DID type entries with transaction number greater than the provided parameter.
```

File: sidetree/lib/core/interfaces/IDidTypeStore.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+   * Inserts or replaces the didtype for a given did.
+   * @param DidTypeModel the did unique suffix and did type
+   */
+  insert (didTypes: DidTypeModel[]): Promise<void>;
+
+  /**
+   * Gets all did unique suffix of the given DID type
+   * @param didType the type of the did
+   */
+  get (didType: string): Promise<string[]>;
```
Comment (thehenrytsai) [inline sidetree/lib/core/interfaces/IDidTypeStore.ts:17]: Are we not concerned about a lot of entries in the returned array? Consider future proofing by making small tweaks:
```suggestion
  /**
   * Gets all did unique suffixes of the given DID type that are greater than `greaterThanTransactionNumber` in transaction time ascending order. Fetches all DIDs of the same DID type if `greaterThanTransactionNumber` is `undefined`'. `hasMore` in returned object is set to `true` if there are too many DIDs to return in on call.
   */
  get (didType: string, greaterThanTransactionNumber?: number): Promise<{did: string[], hasMore: boolean}>;
```

File: sidetree/lib/core/interfaces/IRequestHandler.ts [hunk 10]
```diff
@@ -9,6 +9,11 @@ export default interface IRequestHandler {
    */
   handleOperationRequest (request: Buffer): Promise<ResponseModel>;
 
+  /**
+   * Handles a did type request.
+   */
+  handleDidTypeRequest (request: string): Promise<ResponseModel>;
```
Comment (thehenrytsai) [inline sidetree/lib/core/interfaces/IRequestHandler.ts:15]:
```suggestion
  handleDidTypeRequest (didType: string): Promise<ResponseModel>;
```

File: sidetree/lib/core/models/DidTypeModel.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+
+/**
+ * Did Type Model
+ */
+export default interface DidTypeModel {
+    /** The DID unique suffix. */
+    didUniqueSuffix: string;
+    /** The DID type. */
+    didType: string;
+    /** The transaction number of the transaction this operation was batched within. */
```
Comment (thehenrytsai) [inline sidetree/lib/core/models/DidTypeModel.ts:10]:
```suggestion
    /** The transaction number of the transaction this DID was created in. */
```

**Comment by michaelneale** [comment]
cool stuff @nitro-neal - much needed

**Review by thehenrytsai — COMMENTED** [review]
@nitro-neal, code looks super clean! Just a few minor comments.

The main thing blocking is that I don't actually see any tests written for the new DID Type Store. We really ought to keep the code coverage 100%.

File: sidetree/lib/core/Core.ts [hunk 12]
```diff
@@ -139,6 +144,7 @@ export default class Core {
         await this.transactionStore.initialize();
         await this.unresolvableTransactionStore.initialize();
         await this.operationStore.initialize();
+        await this.didTypeStore.initialize();
         await this.confirmationStore.initialize();
         await this.upgradeDatabaseIfNeeded();
```
Comment (thehenrytsai) [inline sidetree/lib/core/Core.ts:149]: You'll need to update `upgradeDatabaseIfNeeded()` method.

**Review by decentralgabe — APPROVED** [review]
good from me - nice work. would wait for @thehenrytsai 's signoff to merge

**Comment by nitro-neal** [comment]
I have successfully created dids on the testnet and have consumed the didTypes:
image

GET - http://localhost:3000/didtype/0001
```
["EiB28qvpPlmL1dDA1RhqSWPCLQZytp_mVSmrnhnBVUld1Q","EiCs-v7jAyk_cpFfCJlAwgfzRzBkUigNyvkSlzb5ij4QgQ","EiDlfbxQ1-uTgXV_Gq-3iOHeHGpU18Qvxoq_xi1BdrShFg","EiBuJQEvMYNMJmnobbbBTakqy1t-1r_OBkki_tpqGojDHw","EiBu_vRKeMngSTyszG_Nx2xFrwkN735iWclx9nNXWz2zEg","EiCUtwGUL75OPnv1IL2HHq1R5pqYYGv59Uv2Ci3KnMosBg","EiA9KvZ1jPQjKirZ9ya3j97G4n_y3Sx_L-uHVnVQlczHnA","EiC-x1uY4UQApkaunbFmmredz1PhWGFcWfvBOEzBWN5aOw","EiCDPD4_1dhWCNi9cTookWQer7zIan8DkplntjstPPF0uA","EiCDPD4_1dhWCNi9cTookWQer7zIan8DkplntjstPPF0uA"]
```

**Comment by nitro-neal** [comment]
✅ Added New Unit Tests
✅ 100% Code Coverage
✅ Bitcoin testnet integration test

**Comment by csuwildcat** [comment]
This is awesome, I'll pull this into the desktop app very soon.

**Comment by decentralgabe** [comment]
on merge we'll need to get a pipeline up to build a new image and deploy to our node. cc: @tlongwell-block I'll work with you to do this

**Review by thehenrytsai — APPROVED** [review]
`TransactionProcess.ts` in `1.0` and `latest` folder seems not identical and they ought to be.

Otherwise looks good.

**Comment by thehenrytsai** [comment]
> ["EiB28qvpPlmL1dDA1RhqSWPCLQZytp_mVSmrnhnBVUld1Q","EiCs-v7jAyk_cpFfCJlAwgfzRzBkUigNyvkSlzb5ij4QgQ","EiDlfbxQ1-uTgXV_Gq-3iOHeHGpU18Qvxoq_xi1BdrShFg","EiBuJQEvMYNMJmnobbbBTakqy1t-1r_OBkki_tpqGojDHw","EiBu_vRKeMngSTyszG_Nx2xFrwkN735iWclx9nNXWz2zEg","EiCUtwGUL75OPnv1IL2HHq1R5pqYYGv59Uv2Ci3KnMosBg","EiA9KvZ1jPQjKirZ9ya3j97G4n_y3Sx_L-uHVnVQlczHnA","EiC-x1uY4UQApkaunbFmmredz1PhWGFcWfvBOEzBWN5aOw","EiCDPD4_1dhWCNi9cTookWQer7zIan8DkplntjstPPF0uA","EiCDPD4_1dhWCNi9cTookWQer7zIan8DkplntjstPPF0uA"]

There is dup in the last two entries, not sure if you want to dedup.

**Pull request merged**